rfcs/20190729-keras-preprocessing.md

```diff
@@ -0,0 +1,738 @@
+            Default is 'nearest'.
+            Points outside the boundaries of the input are filled
+            according to the given mode:
+            - 'constant': kkkkkkkk|abcd|kkkkkkkk (cval=k)
+            - 'nearest':  aaaaaaaa|abcd|dddddddd
+            - 'reflect':  abcddcba|abcd|dcbaabcd
+            - 'wrap':  abcdabcd|abcd|abcdabcd
+        cval: Float or Int.
+            Value used for points outside the boundaries
+            when `fill_mode = "constant"`.
+        horizontal_flip: Boolean. Randomly flip inputs horizontally.
+        vertical_flip: Boolean. Randomly flip inputs vertically.
+        rescale: rescaling factor. Defaults to None.
+            If None or 0, no rescaling is applied,
+            otherwise we multiply the data by the value provided
+            (after applying all other transformations).
+        preprocessing_function: function that will be applied on each input.
+            The function will run before the image is resized and augmented.
+            The function should take one argument:
+            one image (Numpy tensor with rank 3),
+            and should output a Numpy tensor with the same shape.
+        postprocessing_function: function that will be applied on each input.
+            The function will run after the image is resized and augmented.
+            The function should take one argument:
+            one image (Numpy tensor with rank 3),
+            and should output a Numpy tensor with the same shape.
+        data_format: Image data format,
+            either "channels_first" or "channels_last".
+            "channels_last" mode means that the images should have shape
+            `(samples, height, width, channels)`,
+            "channels_first" mode means that the images should have shape
+            `(samples, channels, height, width)`.
+            It defaults to the `image_data_format` value found in your
+            Keras config file at `~/.keras/keras.json`.
+            If you never set it, then it will be "channels_last".
+        validation_split: Float. Fraction of images reserved for validation
+            (strictly between 0 and 1).
+        interpolation_order: int, order to use for
+            the spline interpolation. Higher is slower.
+        dtype: Dtype to use for the generated arrays.
+    """
+```
+
+```python
+# PROPOSED
+
+def __init__(
+    self,
+    featurewise_center=False,
+    samplewise_center=False,
+    featurewise_std_normalization=False,
+    samplewise_std_normalization=False,
+    zca_whitening=False,
+    rotation_range=0,
+    width_shift_range=0.,
+    height_shift_range=0.,
+    brightness_range=None,
+    shear_range=0.,
+    zoom_range=0.,
+    channel_shift_range=0.,
+    fill_mode='nearest',
+    fill_value=0.,
+    horizontal_flip=False,
+    vertical_flip=False,
+    rescale=None,
+    resize=None,
+    preprocessing_function=None,
+    interpolation_order=1,
+    data_format='channels_last',
+    dtype='float32'):
```

Comment:
This is a lot of constructor args, and it biases us towards adding more with each new feature, rather than finding other ways of setting parameters. Is this the best way to do this, versus having subclasses that handle subsets, or deferring collecting some of these until the method in which they are needed?

Reply:
I guess we can decompose this part in 2 parts: the normalization and the augmentation. The rest is shared. It may add some complexity to what the user needs to do.

Reply:
There is one context in which having lots of arguments is acceptable: when they form a flat list-like configuration, i.e adding more arguments results in linear complexity and maintenance workload, because they don't interact with each other. This is the case here. We're just describing a list of image transformations. Once you have the logic infrastructure to do one image transformation, then doing arbitrarily many does not add much.